intro/intro_tutorial.py

```diff
@@ -0,0 +1,327 @@
+#
+# As a starting example, labeling functions can be based on heuristically **matching keywords**:
+
+# %%
+from snorkel.labeling.lf import labeling_function
+
+@labeling_function()
+def lf_keyword_my(x):
+    """Many spam comments talk about 'my channel', 'my video', etc."""
+    return SPAM if "my" in x.text.lower() else ABSTAIN
+
+
+# %% [markdown]
+# Labeling functions can also use **regular expressions**:
+
+# %%
+import re
+
+@labeling_function()
+def lf_regex_check_out(x):
+    """Spam comments say 'check out my video', 'check it out', etc."""
+    return SPAM if re.search(r"check.*out", x.text, flags=re.I) else ABSTAIN
+
+
+# %% [markdown]
+# Labeling functions can of course flexibly express a range of other **heuristics**:
+
+# %%
+@labeling_function()
+def lf_short_comment(x):
+    """Ham comments are often short, such as 'cool video!'"""
+    return HAM if len(x.text.split()) < 5 else ABSTAIN
+
+
+# %% [markdown]
+# Labeling functions can also use **third-party models**:
+
+# %%
+from textblob import TextBlob
+
+@labeling_function()
+def lf_textblob_polarity(x):
+    """
+    We use a third-party sentiment classification model, TextBlob,
+    combined with the heuristic that ham comments are often positive.
+    """
+    sentiment = TextBlob(x.text).sentiment
+    return HAM if sentiment.polarity > 0.3 else ABSTAIN
+
+
+# %% [markdown]
+# For many more types of labeling functions---including over data modalities beyond text---see the other [tutorials](https://github.com/snorkel-team/snorkel-tutorials) and examples at [snorkel.org](http://snorkel.org).
+# In general the process of developing labeling functions is, like any other development process, an iterative one that takes time- but that, in [many cases](http://snorkel.org), can be orders-of-magnitude faster that hand-labeling training data.
+# For more detail on the process of developing labeling functions and other training data operators in Snorkel, see the [full version of this tutorial](https://github.com/snorkel-team/snorkel-tutorials/tree/master/spam).
+#
+# For now, we proceed by loading the above labeling functions plus some additional similar ones from a utility file:
+
+# %%
+from lfs import lfs
+
+# %% [markdown]
+# ## STEP 2: Combining & Cleaning the Labels
+#
+# Our next step is to apply the labeling functions we wrote to the unlabeled training data; we do this using the `LFApplier` corresponding to our base data class (in this case, the `PandasLFApplier`):
+
+# %%
+from snorkel.labeling.apply import PandasLFApplier
+
+applier = PandasLFApplier(lfs)
+L_train = applier.apply(df_train)
+L_dev = applier.apply(df_dev)
+L_test = applier.apply(df_test)
+
+# %% [markdown]
+# The result of applying the labeling functions (LFs) to the data is (for each split of the data) a sparse _label matrix_ with rows corresponding to data points, and columns corresponding to LFs.
```

Comment:
It's not sparse any more, it's a 2D np.ndarray with -1s for abstains (instead of 0s, which allowed it to be sparse). I guess this needs to be corrected in the original spam tutorial as well.